internal/engine/ruletable/rule_table.go

```diff
@@ -94,27 +102,240 @@ type CelProgram struct {
 		primaryIdx:               make(map[string]map[string]*util.GlobMap[*util.GlobMap[[]*Row]]),
 		scopedResourceIdx:        make(map[string]map[string]*util.GlobMap[struct{}]),
 		log:                      zap.S().Named("ruletable"),
 		schemas:                  make(map[namer.ModuleID]*policyv1.Schemas),
 		meta:                     make(map[namer.ModuleID]*runtimev1.RuleTableMetadata),
 		policyDerivedRoles:       make(map[namer.ModuleID]map[string]*WrappedRunnableDerivedRole),
+		storeQueryRegister:       make(map[namer.ModuleID]bool),
 		scopeMap:                 make(map[string]struct{}),
 		scopeScopePermissions:    make(map[string]policyv1.ScopePermissions),
-		parentRoles:              make(map[string][]string),
-		parentRoleAncestorsCache: make(map[string][]string),
+		parentRoles:              make(map[string]map[string][]string),
+		parentRoleAncestorsCache: make(map[string]map[string][]string),
+		policyLoader:             policyLoader,
+	}
+}
+
+func (rt *RuleTable) LazyLoad(ctx context.Context, resource, policyVer, scope string, inputRoles []string) error {
+	rt.mu.Lock()
+	defer rt.mu.Unlock()
+
+	// A matching scope must have at least one resource or role policy or a mixture of both.
+	// Add rules for policies at all scope levels.
+	// We duplicate rows for all role policy parent roles recursively.
+	//
+	// Rule table resource policies are added as individual units rather than as compilation units.
+	// Therefore, we need to retrieve the compilation unit for each scope, remove all bar the first policy,
+	// and pass all individually to the loadPolicies method.
+	toLoad := []*runtimev1.RunnablePolicySet{}
+
+	// we don't want to add retrieved mod IDs to the cache until the rule table has been successfully updated
+	registryBuffer := make(map[namer.ModuleID]bool)
+
+	// we force lenientScopeSearch when retrieving resource policy sets as lenient scope search is enforced
+	// in the evaluator function. Therefore, to prevent duplicate rows in the rule table, we check the returned
+	// policy scope before adding to `toLoad`
+	addScopedPolicyRules := func(partialScope string) error {
+		var rps *runtimev1.RunnablePolicySet
+
+		cachedPolicyCount := 0
+
+		resourceModID := namer.ResourcePolicyModuleID(resource, policyVer, partialScope)
+
+		// Check to see if the store has already been queried for the given parameters.
+		if policyExists, isQueried := rt.storeQueryRegister[resourceModID]; !isQueried { //nolint:nestif
+			var err error
+			rps, err = rt.getResourcePolicySet(ctx, resource, policyVer, partialScope, true)
+			if err != nil {
+				return err
+			}
+
+			if rps == nil {
+				registryBuffer[resourceModID] = false
+				// we used lenientScopeSearch, so we can assert that no policies exist for all child scopes
+				// TODO(saml) refactor
+				for i := len(scope) - 1; i >= 0; i-- {
+					if scope[i] == '.' || i == 0 {
+						registryBuffer[namer.ResourcePolicyModuleID(resource, policyVer, scope[:i])] = false
+					}
+				}
```

Comment:
This loop gets used a few times in the code so it's probably worth making it an iterator function in the `namer` package and you can use it like `for parent := range namer.ScopeParents(scope) { ... }`.


```go
func ScopeParents(scope string) iter.Seq[string] {
	return func(yield func(string) bool) {
		for i := len(scope) - 1; i >= 0; i-- {
			if scope[i] == '.' || i == 0 {
				if !yield(scope[:i]) {
					return
				}
			}
		}
	}
}
```

Reply:
Cool! I remember looking for a way to do this in the past, didn't realise it had been added!